File: LeetCode/ProblemSet/013.Roman to Integer/roman-to-integer.js
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function (s) {
    let map = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000,
    };
    let special = {
        'IV': 4,
        'IX': 9,
        'XL': 40,
        'XC': 90,
        'CD': 400,
        'CM': 900,
    };

    let r = 0;
    for (let i = 0, len = s.length; i < len; i++) {
        let first = s[i];
        let second = s[i + 1];
        if (second && special[first + second]) {
            r += special[first + second];
            i++;
        } else {
            r += map[first];
        }
    }
    return r;
};

console.log(romanToInt("III"));
console.log(romanToInt("IV"));
console.log(romanToInt("IX"));
console.log(romanToInt("LVIII"));
console.log(romanToInt("MCMXCIV"));

File: LeetCode/Go-Solutions/ProblemSet/0013.Roman_to_Integer/roman-to-integer.go
package _013_Roman_to_Integer

func romanToInt(s string) int {
	m := map[string]int{
		"I":  1,
		"IV": 4,
		"V":  5,
		"IX": 9,
		"X":  10,
		"XL": 40,
		"L":  50,
		"XC": 90,
		"C":  100,
		"CD": 400,
		"D":  500,
		"CM": 900,
		"M":  1000,
	}

	r, a := 0, 0 // r 是结果，a 是每次循环增量
	for i, l := 0, len(s); i < l; i++ {
		a = m[s[i:i+1]]
		if i+1 < l {
			if v, ok := m[s[i:i+2]]; ok {
				a = v
				i++;
			}
		}
		r += a
	}
	return r
}
